### PR TITLE
Applying Suppression Comments to Compound Statements

### PR DESCRIPTION
**Pull request by bnkc**
## Summary

This PR addresses [issue #11216](https://github.com/astral-sh/ruff/issues/11216). It introduces a first attempt at implementing special handling for `# fmt: skip` and `# fmt: off`, so that they apply to entire compound statements. There are still some failing tests to resolve, but I’d appreciate feedback on whether the general approach is sound, as well as any suggestions for improvement.

I’m also curious if this change might require reworking existing tests.

## Test Plan

A new test has been added in `crates/ruff_python_formatter/src/comments/mod.rs`:

```rust
#[test]
fn fmt_skip_applies_to_entire_compound_statement() {
    let source = r#"
if True: x = 0  # fmt: skip
def foo(self, lorem, ipsum, dolor, sit, amet, consectetur, adipiscing, elit, sed, do): ...  # fmt: skip
"#;

    let test_case = CommentsTestCase::from_code(source);
    let comments = test_case.to_comments();

    assert_debug_snapshot!(comments.debug(test_case.source_code));
}
```

The test essentially reproduces the reported issue. There are still edge cases to address, and I’m continuing to refine the solution. I’ve been diving into this codebase for a few days, so apologies for any oversights. Looking forward to your feedback!

### CODE REVIEW

**Review by MichaReiser — COMMENTED** [review]
Solving this inside the comment placement is clever. I like it. 

My only concern is that not all clause headers have their own node. For example, the `else` header of many statements is just a `Vec<Node>`:

```
while test:
    pass
else:  ... # fmt: skip
```

I recommend to look into how to support these clause headers as a next step.

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 1]
```diff
@@ -415,6 +422,69 @@ fn handle_end_of_line_comment_around_body<'a>(
...
+            return CommentPlacement::trailing(preceding_node, comment);
+        }
+    }
+
+    // Handle special cases like `else`, `except`, `finally` blocks that are not their own nodes
+    if let Some(parent) = comment.enclosing_parent() {
+        if is_clause_header(parent) {
+            // Apply fmt: skip if the comment is on the same line as the 'else', 'except', or 'finally'
+            let header_line_start = locator.line_start(parent.start());
+            let comment_line_start = locator.line_start(comment.start());
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:477]: I think this is an over-simplification. E.g. 

```python
def test(
	a, 
	b, 
): ... # fmt: skip

def test() -> List[
	str,
	int,
]: ... # fmt: skip

if (
	a and b and c and [
		1, 
		2,
): ... # fmt: skip 
```

There's `ClauseHeader` that has some fairly complicated logic to find the exact range of the header

https://github.com/astral-sh/ruff/blob/138e70bd5c01d61061247c43b1bbb33d0290a18f/crates/ruff_python_formatter/src/statement/clause.rs#L40-L70

maybe we can reuse some of the logic here

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 2]
```diff
@@ -415,6 +422,69 @@ fn handle_end_of_line_comment_around_body<'a>(
...
+
+            if header_line_start == comment_line_start {
+                return CommentPlacement::trailing(parent, comment);
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:480]: I'm not sure if making the comment a trailing comment is correct. Doesn't it suppress formatting for the entire node? I think the comment should be a dangling comment, see 

https://github.com/astral-sh/ruff/blob/6ac61d7b89b2ac3789c2d8ecde505f8c739c6f35/crates/ruff_python_formatter/src/comments/placement.rs#L373-L392

**Comment by MichaReiser** [comment]
I looked at this a bit more closely and I don't think a `placement.rs` only solution is going to work. What makes me belive so is that I changed the source in `quick_test` to 

```rust
let source = r#"
def foo(self, lorem, ipsum, dolor, sit, amet, consectetur, adipiscing, elit, sed, do): print( " teest ")  # fmt: skip
"#;
```

The test then fails with:

```
The node has dangling comments that need to be formatted manually. Add the special dangling comments handling to `fmt_fields`.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Which makes sense when we take a look at the `FormatClauseHeader` implementation

https://github.com/astral-sh/ruff/blob/a0afe7857eb26554da2694422d536b14479b115a/crates/ruff_python_formatter/src/statement/clause.rs#L368-L395

Specifically at how it formats a suppressed clause header

https://github.com/astral-sh/ruff/blob/a0afe7857eb26554da2694422d536b14479b115a/crates/ruff_python_formatter/src/verbatim.rs#L954-L980

You can see how it only formats the clause header in verbatim mode but not the body. We need to extend the clause handling with a branch that skips both the clause header and body formatting when the sinlge-body statement is suppressed

**Comment by bnkc** [comment]
@MichaReiser Thanks for the feedback! I'll get to this over the weekend :)

**Pull request closed without merging**